apply_neighbourhood.json

```diff
@@ -0,0 +1,74 @@
+    "categories": [
+        "cubes"
+    ],
+    "parameters": [
+        {
+            "name": "data",
+            "description": "A data cube.",
+            "schema": {
+                "type": "object",
+                "subtype": "raster-cube"
+            }
+        },
+        {
+            "name": "process",
+            "description": "Process to be applied on all neighbourhoods. The specified process needs to accept an array as parameter and must return an array with least one element. A process may consist of multiple sub-processes.",
+            "schema": {
+                "type": "object",
+                "subtype": "process-graph",
+                "parameters": [
+                    {
+                        "name": "data",
+                        "description": "A labeled array with elements of any type.",
+                        "schema": {
+                            "type": "array",
+                            "subtype": "labeled-array",
+                            "items": {
+                                "description": "Any data type."
+                            }
+                        }
+                    },
+                    {
+                        "name": "context",
+                        "description": "Additional data passed by the user.",
+                        "schema": {
+                            "description": "Any data type."
+                        },
+                        "optional": true,
+                        "default": null
+                    }
+                ]
+            }
+        },
+        {
+            "name": "size",
+            "description": "Neighbourhood sizes along each dimension. Default is to provide only one value.\n This object maps dimension names to either a physical measure, (e.g. 100 m, 10 days), or pixels (e.g. 32 pixels).\nTo include all values along a dimension, use `all`. This may result in an exception if the dimension is too large to be fully included.",
+            "schema": {
+                "type": "object"
+            }
+        },
+        {
+            "name": "overlap",
+            "description": "Overlap of chunks along each dimension. This allows neighbourhoods to overlap. For instance a temporal dimension can add 1 month before and after a neighbourhood. In the spatial dimensions, this is often a number of pixels. Modifying overlapping data in subsequent operations should have no effect.",
+            "schema": {
+                "type": "object"
+            }
+        }
+    ],
+    "returns": {
+        "description": "A data cube with the newly computed values. The resolution and the number of dimensions are the same as for the original data cube.",
```

Comment:
> the resolution .. are the same as for the original data cube

Will this be the case in general? If you use a large `size` (and small `overlap`), the result will have a "downsampled" resolution if I understand correctly.

Reply:
Maybe I misunderstood the operation, but I thought the resolution will not change.

Reply:
No spatial resolution should remain unchanged. For bands and time dimensions, our backend could be more flexible, but not sure if that is a general thing. (People can always set all values to nodata to 'remove' something as well).
So it's more like apply_dimension, which can change timestamps in the temporal dimension if I'm correct.

Reply:
I guess than the description in the process parameter:
> The specified process needs to accept an array as parameter and must return an array with least one element. A process may consist of multiple sub-processes.

has to be removed

Reply:
Ok I misunderstood the goal of this process apparently. 
So in terms of resolution, input/output cardinality it's a generalization of a convolution (`apply_kernel`)?

Reply:
As far as I understood, yes.

Reply:
but then I don't understand why you have both `size` and `overlap`

Reply:
I understood it like that:
The overlap is just additional information you may use for computation, but once you "replace" the data in the original data with the newly computed data, the overlap is skipped.

Reply:
As far as I understand focal operations (and convolutions), the window size and window overlap are tied to each other 1-on-1. 
For example in 1 dimension: if you process the value at index `x`, you take input values in the window `[x-n,x+n]` (so the window size is `2n+1`). For the next pixel at index `x+1`, the shifted window has an overlap of `2n` pixels  with the previous window, so overlap size = window size - 1 pixel.

If overlap is about how much margin should be taken when doing tile based processing, then still the overlap can be derived from the window size I would think (`n` in the 1D example).

Reply:
@jdries will add a better description to the process, including the ML example.

Reply:
for completeness, to conclude this subthread: as discussed in telco this morning: `size` is not only (part of) the size of the input window, but also the size of the block of output pixels you write for a given position of the sliding window (e.g. with convolutions this is always 1 pixel, but `apply_neighbourhood` allows larger areas at once). The sliding window moves with multiples of `size`. And `overlap` is an additional extension of the sliding window to take more input pixels (output "block" size and step size are  still `size`).

Reply:
Was adding something similar to this conclusion also part of your commit https://github.com/Open-EO/openeo-processes/pull/163/commits/9a2cc6cf2b1758ebcedcfb90df1048cbd644971a, @jdries ?

Reply:
@soxofaan I used some of your wording to improve the process description.